pkg/debuginfo/store.go

```diff
@@ -313,19 +358,43 @@ func (s *Store) Symbolize(ctx context.Context, m *pb.Mapping, locations ...*meta
 		if err != nil {
 			return nil, fmt.Errorf("failed to fetch: %w", err)
 		}
+		downloadedFromDebugInfod = true
 	}
 
 	// Let's make sure we have the best version of the debug file.
-	hasDWARF, err := elfutils.HasDWARF(objFile)
-	if err != nil {
-		return nil, fmt.Errorf("failed to check for DWARF: %w", err)
+	if err := elfutils.ValidateFile(objFile); err != nil {
+		level.Warn(logger).Log("msg", "failed to validate debug information", "err", err)
+		// Mark the file as corrupted, and let the client try to upload it again.
+		err := s.metadataManager.corrupted(ctx, buildID)
+		if err != nil {
+			level.Warn(logger).Log(
+				"msg", "failed to mar debug information",
+				"err", fmt.Errorf("failed to update metadata for corrupted: %w", err),
+			)
+		}
+		if !downloadedFromDebugInfod {
```

Comment:
This looks like a great candidate for another metadata field! We can do this in another PR, but I think it would help to debug gnarly situations, as well as potentially making the code easier to follow

Reply:
This is a to-do item that I have on the agent side. I think we should add a source type enum file to metadata and the API. So that this could help us debug issues. And we can improve the state machine we have in hear.

Reply:
What do you think, shall we just add it as part of this PR?

Reply:
Either works, but this PR is already quite comprehensive, should we do it in another PR?

Reply:
Created an issue for it https://github.com/parca-dev/parca/issues/1258